### PR TITLE
Use lxml instead of BeautifulSoup for text extraction (Fix #232, Fix wellcometrust/datalabs#404)

### LINKED ISSUE #232: Investigate pulling text directly from PDFs instead of via poppler -> HTML
As @jdu noted on https://github.com/wellcometrust/policytool/issues/227, we'd do well to look into working directly from PDFs instead of the current PDF -> HTML -> BeautifulSoup route.

Scope for a 3-5 day investigation, I'd think..?

### PR DESCRIPTION
**Pull request by ivyleavedtoadflax**
# Description

Following discussion in https://github.com/wellcometrust/datalabs/issues/404 and some exploration in https://github.com/wellcometrust/datalabs/pull/417 this PR switches the policytool from using `BeautifulSoup` to `lxml`.

The result is an improvement in the way that references are extracted from pdfs measured using the benchmarks developed by @lizgzil, and by manual examination of the extracted data.

Note that this is a significant change to the way that the text extractor works, and so there are some differences in the way the module performs.:

1. Tests relating to **bold** text fail completely, and for this reason I have commented them out. This PR makes no attempt to recognise **bold** text at all. My reason for this is that whilst the `policytool.pdf_parse.tools.extraction._find_elements()` function seems to use the bold attribute, I could not find anywhere in `policytool.pdf_parse.pdf_parse` where the `bold` attribute is actually set to `True`. Unless I have missed something, it is always set to `False`. Happy to be corrected on this; but if the direction of the reference section finder is moving towards ML, then the `bold` attribute is probably may not be very important.
1. Some tests relating to text size return slightly different results, for which I have amended the tests. 

## Type of change

Please delete options that are not relevant.

- [x] :bug: Bug fix

# How Has This Been Tested?

```
make test
make # docker tests
```

[Automated evaluation](https://github.com/wellcometrust/policytool/pull/243#issuecomment-538379232) (higher is better):

|Metric|Old|This PR|
|---|---|---|
|Score (1- Mean normalised Levenshtein distance)|0.388|0.56|
|Lenient accuracy (micro) (normalised Levenshtein < 0.3)|0.308|0.5|

[Manual evaluation](https://github.com/wellcometrust/policytool/pull/243#issuecomment-538379232):

|Description|n|
|---|---|
|Good matches|7|
|Good match with additional text|7|
|Some missing references|5|
|Total failure|3|
|Excluded due to issues in Actual|1|

Lots of eyeballing :eyes:.

# Checklist:

- [x] My code follows the style guidelines of this project (pep8 AND pyflakes)
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I included tests in my PR
- [x] New and existing unit tests pass locally with my changes
- [x] If my PR aims to fix an issue, I referenced it using `#(issue)`

### CODE REVIEW

**Comment by nsorros** [comment]
Do the tests pass? Is it possible to test the effect this change has on the evaluation data and tests that liz has written?

**Comment by nsorros** [comment]
@SamDepardieu can we get your input on this?

**Comment by nsorros** [comment]
Don't forget to mention all the issues that will close if we merge this.

**Comment by SamDepardieu** [comment]
I really like what I'm looking at 👍 
Why is this still a draft? Do you plan on changing more?

**Comment by ivyleavedtoadflax** [comment]
[Use lxml in text extractor](https://trello.com/c/i4yvB4zL/838-use-lxml-in-text-extractor)

**Comment by ivyleavedtoadflax** [comment]
Currently working through test failures

> I really like what I'm looking at 
> Why is this still a draft? Do you plan on changing more?

Yep, working on getting tests to pass now :+1:

**Comment by ivyleavedtoadflax** [comment]
@SamDepardieu @hblanks quick question, having some issues getting tests to pass, and I just want to check this isn't why:

In `./policytool/pdf_parser/tests/test_pdf_objects.py` the test file used is described as:

```
"""Test file content (html transcription):
<h1>Test</h1>
Test
<b>Test bold</b>
<h1>References</h1>
<ol>
    <li>Test</li>
    <li>Test</li>
    <li>Test</li>
</ol>"""
```
(This file is imported with `from policytool.scraper.tests.common import TEST_PDF`). However `./polictool/pdf_parse/pdf_parse.py` runs the `pdftohtml` tool in the following way:

```
        # Run pdftohtml on the document, and output an xml formated document
        cmd = [
            'pdftohtml',
            '-i',
            '-xml',
            '-zoom',
            '1.5',
            document.name,
            tf.name
        ]
```
The `-xml` flag outputs an xml document of the format:

```
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE pdf2xml SYSTEM "pdf2xml.dtd">

<pdf2xml producer="poppler" version="0.62.0">
<page number="1" position="absolute" top="0" left="0" height="1262" width="892">
	<fontspec id="0" size="12" family="Times" color="#000000"/>
	<fontspec id="1" size="25" family="Times" color="#000000"/>
	<fontspec id="2" size="19" family="Times" color="#000000"/>
	<fontspec id="3" size="11" family="Times" color="#000000"/>
<text top="1179" left="108" width="683" height="14" font="0">quality assurance guidelines for medical physics services 2nd edition may 2005 updated dec 2018.doc  </text>
<text top="141" left="446" width="8" height="25" font="1"><b> </b></text>
...
<text top="430" left="446" width="8" height="25" font="1"><b> </b></text>
<text top="462" left="212" width="475" height="25" font="1"><b>QUALITY ASSURANCE GUIDELINES </b></text>
<text top="494" left="218" width="472" height="25" font="1"><b>FOR MEDICAL PHYSICS SERVICES  </b></text>
<text top="526" left="446" width="8" height="25" font="1"><b> </b></text>
<text top="556" left="220" width="463" height="19" font="2"><b>National Breast Screening Quality Assurance  </b></text>
<text top="581" left="287" width="324" height="19" font="2"><b>Coordinating Group for Physics </b></text>
```

Is just an error in the docs in `./policytool/pdf_parser/tests/test_pdf_objects.py`? I tried to read `TEST_PDF` in python to check, but I can't seem to decode it:

```
In [7]: from policytool.scraper.tests.common import TEST_PDF

In [8]: with open(TEST_PDF, 'rb') as fb:
   ...:     foo = fb.read()
   ...: 

In [9]: foo.decode("utf-8")
---------------------------------------------------------------------------
UnicodeDecodeError                        Traceback (most recent call last)
<ipython-input-9-f1b958a095b8> in <module>
----> 1 foo.decode("utf-8")

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xb5 in position 11: invalid start byte
```

**Comment by ivyleavedtoadflax** [comment]
```Date of evaluation = 2019-10-04-0927

-----How well the scraper finds the references section:-----

Score
0.56

Lenient F1-score (references section exists or not)
0.947

Metrics by provider
Provider                                         msf    nice  unicef  who_iris        all
Number of pdfs included                       15.000  17.000  31.000    32.000  95.000000
Number of pdfs with sections text              7.000   5.000   9.000     5.000  26.000000
Proportion of pdfs with sections text          0.467   0.294   0.290     0.156   0.274000
Lenient F1 score for all sections included     0.933   0.922   0.925     0.990   0.947000
Number of pdfs with a bibliograph section      1.000   0.000   0.000     1.000   2.000000
Number of pdfs with a endnotes section         2.000   0.000   4.000     0.000   6.000000
Number of pdfs with a reference section        4.000   5.000   5.000     4.000  18.000000
Proportion with a bibliograph section          0.067   0.000   0.000     0.031        NaN
Proportion with a endnotes section             0.133   0.000   0.129     0.000        NaN
Proportion with a reference section            0.267   0.294   0.161     0.125        NaN
Lenient F1 score for the bibliograph section   0.933   1.000   1.000     1.000   0.989474
Lenient F1 score for the endnotes section      1.000   1.000   1.000     1.000   1.000000
Lenient F1 score for the reference section     0.867   0.765   0.774     0.969   0.852632

Number of pdfs with sections text
26

Mean normalised Levenshtein distance
0.44

Strict accuracy (micro)
0.0

Lenient accuracy (micro) (normalised Levenshtein < 0.3)
0.5

Mean normalised Levenshtein distance for the reference section
0.489

Strict accuracy for the reference section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the reference section
0.5

Mean normalised Levenshtein distance for the endnotes section
0.272

Strict accuracy for the endnotes section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the endnotes section
0.5

Mean normalised Levenshtein distance for the bibliograph section
0.498

Strict accuracy for the bibliograph section
0.0

Lenient accuracy (normalised Levenshtein< 0.3) for the bibliograph section
0.5

-----How well the splitter predicted how many references there were:-----

Score
0.757

Number of reference sections in sample
70

Proportion below threshold of 40
0.757

Median difference
20.0

Metrics grouped by source
Source                                     msf   nice  unicef  who_iris
Number of reference sections in sample   9.000  25.00    14.0    22.000
Median difference                       33.300  16.70    37.1    13.900
Proportion below threshold of 40         0.667   0.92     0.5     0.773

-----How well the parser predicted reference component texts:-----

Score
0.862

Number of references in sample
205

Number of non-blank reference components
935

Strict accuracy (micro)
0.403

Lenient accuracy (micro)(normalised Levenshtein < 0.3)
0.817

Mean normalised Levenshtein distance (all categories)
0.138

Metrics per category
                                                Authors  Journal  ...    Title  PubYear
Number of non-blank reference components        176.000  154.000  ...  202.000  188.000
Strict accuracy                                   0.341    0.337  ...    0.244    0.293
Lenient accuracy(normalised Levenshtein < 0.3)    0.615    0.839  ...    0.610    0.956
Mean normalised Levenshtein distance              0.299    0.137  ...    0.256    0.036

[4 rows x 7 columns]

-----How well the matcher matched references:-----

Score
0.944

Number of references in sample
2000

Micro average F1-score
0.944

Micro average recall
0.944

Micro average precision
0.944

Classification report
              precision    recall  f1-score   support

       False       0.90      1.00      0.95      1000
        True       1.00      0.89      0.94      1000

    accuracy                           0.94      2000
   macro avg       0.95      0.94      0.94      2000
weighted avg       0.95      0.94      0.94      2000


Frequency table of match types
What did we expect?    Match expected  No match expected
What did we find?                                       
Correct match found               890                  0
Incorrect match found               0                  2
No match found                    110                998
```

**Comment by ivyleavedtoadflax** [comment]
Manual evaluation of test documents used in automated analysis:

|file|actual size|predicted size|levenshtein distance|comment|
|---|---|---|---|---|
|6f6bf7867b9b35f2dff5f3d8583718e5|5929|234328|0.97|Captures all references but overshoots end of the reference section.|
|ad6a1c16d161c6989d2d7a2d793b5c0a|8014|493|0.95|Ref section finder failed to find the correct section.|
|a497d288330b0ff0f308c43c93417a7b|33126|30589|0.32|Start of reference section missing. End of reference section overshot|
|a3aca97e2bae1463a99cb65bda638345|26178|20649|0.49|Missed beginning of reference section, undershoots end.|
|e049d3f401020d82f37c47425062e6d4|3172|1209|0.89|Model found 'references' from the contents page, and not the actual reference section.|
|05582252f4a6ba2a1f4201ef9eefcab9|46970|108847|0.7|Misses start of reference section, undershoots end of reference section.|
|43715f478bd1621b8f8aecf2ac693ea4|16521|18125|0.3|Start of reference section missing. End of reference section overshot|
|a02fd56180bade68255845167a9e6910|8321|11504|0.28|End of reference section overshot. All references captured.|
|c5444dde13b2a88e91d2341de1870fef|1997|4033|0.5|All references found, overshoots the end of the reference section.|
|28ff5ef02809e86d99c8c25c01b256d9|5460|1495|0.84|Predicted fails completely and extracts elements from the contents page.|
|85ceca8229a0c7b9a6e11e39c72e33ef|3783|1793|0.6|Predicted misses start of reference section.|
|b687b08bd4d460de3c7f634462159176|5936|16190|0.63|All references found, predicted overshoots the end of the reference section.|
|c0922daab9ad3cab2128d929aee54d2e|25998|34984|0.26|All references captured, predicted overshoots the end of the reference section|
|9fc4cfe01187b5a55d8bc0aa3b8bb017|21790|26725|0.2|Both predicted and actual show issues relating to reference numbers. Probably from dual columns. Predicted seems to capture more references.|
|e7c7cbe41901d3764f5bec0eb8d81a60|16919|18727|0.11|Predicted overshoots the end of the reference section. Some references found in predicted not in actual.|
|b34cab0584160c38c6b67ffcc20a36db|6212|6349|0.03|Small formatting differences|
|4e0b8316bfb6e526b9d5c45f074b0227|9325|9800|0.05|Almost identical. Predicted text includes some footer text.|
|687751f9d137b7940a702111e4e25d3a|10208|10493|0.03|Almost identical. Some minor formatting differences.|
|4900e8cec4985b9ed8e4e27d239d1005|9212|9595|0.04|Very close. Some additional vertical text has been captured in the predicted.|
|768fcec7ea4c9ec12f1959e885a6d1ba|10007|10406|0.04|Minor formatting differences.|
|0e2bbb2d8a662cce8ea1d6f696db4c2e|29033|29540|0.02|Minor formatting differences.|
|18acd3e92334926c1b82aa19f2a7eb1c|4268|4679|0.09|End of reference section overshot. All references captured.|
|d4a26bff8ba39359d99eb3bdb72cc48d|26112|28855|0.1|Formatting differences|

**Comment by ivyleavedtoadflax** [comment]
@hblanks @SamDepardieu I think this is good to go from a data science perspective.

File: policytool/pdf_parser/tests/test_pdf_objects.py [hunk 1]
```diff
@@ -3,7 +3,7 @@
 
 from policytool.pdf_parser.pdf_parse import parse_pdf_document
 from policytool.pdf_parser.objects.PdfObjects import PdfFile
-from policytool.scraper.tests.common import TEST_PDF
+from policytool.scraper.tests.common import TEST_PDF, TEST_PDF_MULTIPAGE
```
Comment (SamDepardieu) [inline policytool/pdf_parser/tests/test_pdf_objects.py]: That's a nice addition

File: policytool/pdf_parser/tests/test_pdf_objects.py [hunk 2]
```diff
@@ -97,19 +96,19 @@ def test_lines_by_keyword(self):
     def test_lines_by_keywords(self):
         keywords = ['bold', 'test', 'machine']
         keyword_lines = self.pdf_file_object.get_lines_by_keywords(keywords)
-        self.assertTrue('bold' in keyword_lines.keys())
-        self.assertEqual(len(keyword_lines['bold']), 1)
+        #self.assertTrue('bold' in keyword_lines.keys())
+        #self.assertEqual(len(keyword_lines['bold']), 1)
```
Comment (SamDepardieu) [inline policytool/pdf_parser/tests/test_pdf_objects.py]: So, my legacy (and not very reliable) way of finding reference sections worked a bit better using bold text to find titles. Do you have similar accuracy results even without using that?

File: policytool/refparse/algo_evaluation/compare_found_sections.py [hunk 3]
```diff
@@ -0,0 +1,45 @@
+# coding: utf-8
+"""
+Utility for comparing predicted and actual reference sections.
+
+Takes the scarpe_data.csv produced by evaluate_algo.py, and produces an
+interactive dashboard through which actual and predicted references sections
+can be compared.
+
+Requires streamlit>=0.47.3
```
Comment (SamDepardieu) [inline policytool/refparse/algo_evaluation/compare_found_sections.py]: What does this library do?

File: policytool/scraper/tests/common.py [hunk 4]
```diff
@@ -8,3 +8,5 @@ def get_path(p):
     )
 
 TEST_PDF = get_path('pdfs/test_pdf.pdf')
+TEST_PDF_MULTIPAGE = get_path('pdfs/test_pdf_multipage.pdf')
```
Comment (SamDepardieu) [inline policytool/scraper/tests/common.py]: Did you add this PDF to the repo? I can't find it

Comment (ivyleavedtoadflax) [reply policytool/scraper/tests/common.py]: Good spot - I've added it in 68b3f2c

File: policytool/pdf_parser/tests/test_pdf_objects.py (hunk 1 above)
Comment (ivyleavedtoadflax) [reply policytool/pdf_parser/tests/test_pdf_objects.py]: :+1:

File: policytool/refparse/algo_evaluation/compare_found_sections.py (hunk 3 above)
Comment (ivyleavedtoadflax) [reply policytool/refparse/algo_evaluation/compare_found_sections.py]: Streamlist creates an interactive visualisation that is really useful for comparing predicted and actual reference sections. It really is not critical to the functioning of Reach, nor is this a script that I think we would need to run very often, so for this reason I have not added it to the requirements, but instead included a note on it here.

File: policytool/pdf_parser/tests/test_pdf_objects.py (hunk 2 above)
Comment (ivyleavedtoadflax) [reply policytool/pdf_parser/tests/test_pdf_objects.py]: So this is a good question.

Firstly, yes, the accuracy is better using this method than the BeautifulSoup method, according to @lizgzil's [metrics](https://github.com/wellcometrust/policytool/pull/243#issuecomment-538379232). 

It is not perfect though, as you can see in a [manual evaluation](https://github.com/wellcometrust/policytool/pull/243#issuecomment-538531917), but as we discussed briefly on slack, I think the real solution to this is either:

* an ML solution to recognising reference sections - or 
* the implementation of the neural reference splitter to which in principle we can just pass the whole document.

Secondly, from what I saw in the code (and please do correct me if I'm wrong), I don't think the bold attribute is actually being used in the existing codebase.

There are only two mentions of the `PdfLines` class in which bold text is recorded: https://github.com/wellcometrust/policytool/search?q=PdfLine&unscoped_q=PdfLine

In the latter where it is actually instantiated, it is always set to `False`: https://github.com/wellcometrust/policytool/blob/8d405645d50602ca2d6c96e99ca146008f2cbb4d/policytool/pdf_parser/pdf_parse.py#L99:L104.

If this is right, then `_find_elements()` which is called by `grab_section()` will always default to the [following](https://github.com/wellcometrust/policytool/blob/8fdf61e6e5aaf7a80b5887fdcc4aad4044dfe347/policytool/pdf_parser/tools/extraction.py#L47:L55), and so it does not matter whether we record bold text or not.

```
    for fsize in list_fonts:
        for line in pdf_file.get_lines_by_font_size(fsize):

            # If a font is bold, in title font and bigger than other,
            # it is probably a title
            text = line.text

            # PdfFile has some bold font
            if pdf_file.has_bold:
            ...

            # PdfFile has been parsed using pdftotext or has no bold
            else:
                ...
```

Comment (SamDepardieu) [reply policytool/pdf_parser/tests/test_pdf_objects.py]: It is used in `_find_elements()`, precisely, in the extremly unclear (my bad) title extracting part:

```
for fsize in list_fonts:
        for line in pdf_file.get_lines_by_font_size(fsize):

            # If a font is bold, in title font and bigger than other,
            # it is probably a title
            text = line.text

            # PdfFile has some bold font
            if pdf_file.has_bold:
                font_is_bigger = fsize >= (upper_mean
                                           + (upper_mean - mean_fonts))

                **font_is_big_and_bold = font_is_bigger and line.bold**

                font_is_title_like = (fsize > upper_mean + 2
                                      and line.font_face == max_fonts_name)

                if font_is_big_and_bold or font_is_title_like:
                    match = re.search(regex, text, re.IGNORECASE)
                    if match:
                        titles_font_size = line.size
                        break

            # PdfFile has been parsed using pdftotext or has no bold
            else:
                font_is_bigger = fsize > mean_fonts + 1
                font_is_title_like = (fsize > upper_mean + 2
                                      and line.font_face == max_fonts_name)
                if font_is_bigger or font_is_title_like:
                    match = re.search(regex, text, re.IGNORECASE)
                    if match:
                        titles_font_size = line.size
                        break
```
The main purpose was to build a score letting us identify more easily the font size used by the title, and thus the titles themselves.

Although it doesn't really matter if your upgrade achieves better performances, and as you said, the only future I see for this part of the pipeline is either ML or NN, but certainly not anything close to what it is at the moment 👍

Comment (SamDepardieu) [reply policytool/pdf_parser/tests/test_pdf_objects.py]: (Also you're right, at the moment it's always set to False anyway)

**Comment by SamDepardieu** [comment]
@ivyleavedtoadflax I'n happy for you to merge this whenever you want ;)

**Pull request merged**